### ISSUE TITLE
GO ids are reused / not unique

### ISSUE THREAD

**Issue by 8bitskull** [defold/defold#9851]
This can lead to bugs that are very difficult to debug. For example:

* I create an object in a manager script, and set up a timer to delete this object.
* I unload the current collection proxy and load in a new one. The object is deleted, but I forget to remove the timer.
* The timer fires, and deletes an object with the original id - which is now potentially a completely different object.

Note that this setup was poorly constructed by me in the first place. Because I neglected to reset the timer on collection switch, I _was_ often getting an error about game objects not existing when a deletion request came up. So in an attempt to fix this, I wrapped go.delete() in a go.exists() check first. Inadvertently, this means I created a situation that worked fine most of the time, but when I bumped into reused game object ids, I experienced this very tricky debug situation.

**Comment by aglitchman** [comment]
But it's not a bug, that's how it works by design, because the objects are taken from the pool. Right?

**Comment by AGulev** [comment]
> But it's not a bug, that's how it works by design, because the objects are taken from the pool. Right?

Yes, but maybe it's possible to do something with that e.g. creating new unique id every time it's taken from pool, but it might be a problem in debug builds where hashes has cached strings to be able to debug it. 
So it should be investigated, but I agree it's a reason of very rare but extremely annoying bug which hard to catch and debug.

**Comment by redoak** [comment]
Perhaps ids and indices can be implemented like what's usually done in ECS.

Generally, an entity id is made up of two integers: an index and a generation. The index is the (array) storage location of the entity. The generation's purpose is to avoid issues like 8bitskull's. It's incremented each time the id is recycled, allowing index reuse while making the id unique. This is generally implemented by packing both values into a single int (the id), e.g. 24 bits for the index and 8 bits for the generation.

In the engine, I see there's a 32 bit instance index, which is pulled and pushed from an index pool, and assigned to game objects. 

Without having looked too close at this, one solution could be to turn the index pool into an (integer) id pool. When an id is pushed back to the pool, its generation would be incremented before being put on the internal stack. Everywhere where an id (what's currently referred to as index in the code) is used to do an array lookup or similar, the generation part of the id would need to be stripped to get the index. This can be done with bitwise and: `int index = id & INDEX_MASK`.

Regarding the string/hash id, it will be less readable since the number will get very big (and the string longer) as a result of the bit packing. Maybe this is an issue, maybe not.
